ls1intum/Artemis
# Enhancement/tutor dashboard/support late submissions

### Checklist

### Motivation and Context

Resolves #1376

The tutor dashboard (and other pages) do not distinguish between in-time submissions and late submissions for exercises. 
While this can lead to confusion or a lack of clarity for tutors, it also introduces statistical errors on certain calculations (see issue #1376)

### Description



1. Introduced separate variables for in-time and late submissions, as well as assessment counts of in-time and late submissions. In both java and js the class *DueDateStat* is used to give some structure. It contains an **in-time** value and a **late** value. In js, it also contains a **total** computed value.  
   Changes are made mainly to: 
  - [Java Models]: +DueDateStat.java, Exercise.java, StatsForInstructorDashboardDTO.java
  - [Java Api]: CourseResource.java, ExerciseResource.java
  - [JS]: +due-date-stat.model.ts, instructor-course-dashboard, tutor-course-dashboard, tutor-exercise-dashboard

 2. Extended the statistics table to properly show and calculate late submission stats in tutor-exercise-dashboard and instructor-exercise-dashboard.

 3. Extended the *tutor-participation-graph* by a separate progress bar for late submissions / assessments. This is shown parallel to the in-time bar and only, if there are any late submissions. The overall progress status (coloring) is not affected by the late submission status.

 4. Modify the assessment order, so that late-submissions are assessed last and only if there are no in-time submissions left to assess.
    - The Modeling Exercises have a Compass Service, which, when requesting a new assessment, calculates the submission with the most knowledge gain for automatic assessments. Currently this is done **before** doing the prioritization. Therefore it can sill happen that late submissions are selected for new assessments while there are unassessed in-time submissions left. 

### Steps for Testing


1. Log in as different students
2. Create in-time as well as late submissions for an exercise
3. Log in as a tutor
4. Choose a course with late submissions
5. Navigate to Course Administration -> Tutor Dashboard ( -> Exercise Dashboard )
6. Review: participation graph and statistics table
7. Check correct choosing of submissions for new assessments
8. Navigate to Course Administration -> Instructor Dashboard
9. Review: statistics table and exercises list

### Screenshots



<img width="419" alt="Participation Graph" src="https://user-images.githubusercontent.com/13920539/82185561-ae761e80-98e9-11ea-9a67-3487fa1dad64.png">
<img width="565" alt="Bildschirmfoto 2020-05-18 um 16 54 37" src="https://user-images.githubusercontent.com/13920539/82227585-4692f880-9928-11ea-86d1-0c9e03782404.png">
